backend/apis/leads/list-leads.api.js

```diff
@@ -152,71 +147,130 @@ app.get("/api/leads", (req, res, next) => {
 
     const totalCount = totalCountRows[0]["FOUND_ROWS()"];
 
-    const mapLeadsData = leadRows.map(result => {
-      const leadServices = JSON.parse(result.leadServices);
-      const eventSources = JSON.parse(result.eventSources);
-
-      return {
-        id: result.leadId,
-        dateOfSignUp: responseDateWithoutTime(result.dateOfSignUp),
-        leadStatus: result.leadStatus === null ? "active" : result.leadStatus,
-        contactStage: {
-          first: result.firstContactAttempt,
-          second: result.secondContactAttempt,
-          third: result.thirdContactAttempt
-        },
-        inactivityReason: result.inactivityReason,
-        eventSources: eventSources.map(event => ({
-          eventId: event.eventId,
-          eventName: event.eventName,
-          eventLocation: event.eventLocation,
-          eventDate: event.eventDate
-        })),
-        firstName: result.firstName,
-        lastName: result.lastName,
-        fullName: responseFullName(result.firstName, result.lastName),
-        phone: result.phone,
-        smsConsent: responseBoolean(result.smsConsent),
-        zip: result.zip,
-        age: result.age,
-        gender: result.gender,
-        leadServices: leadServices.map(service => ({
-          id: service.serviceId,
-          serviceName: service.serviceName
-        })),
-        clientId: result.clientId,
-        isDeleted: responseBoolean(result.isDeleted),
-        createdBy: {
-          userId: result.addedBy,
-          firstName: result.createdByFirstName,
-          lastName: result.createdByLastName,
-          fullName: responseFullName(
-            result.createdByFirstName,
-            result.createdByLastName
-          ),
-          timestamp: result.dateAdded
-        },
-        lastUpdatedBy: {
-          userId: result.modifiedBy,
-          firstName: result.modifiedByFirstName,
-          lastName: result.modifiedByLastName,
+    const leads = [];
+
+    for (let i = 0; i < leadRows.length; i++) {
+      const getMoreData = mysql.format(
+        `
+        SELECT
```

Comment:
Two database trips instead of one is not ideal for performance. Nothing is immediately coming to mind as a solution to avoid the two trips, though. Thoughts on how?

Reply:
I just realized that this isn't two database trips, it is `n + 1` database trips (one trip to get the list of leads and one more **per lead**.

This is an absolute showstopper for this PR. We should never be making 101 trips to the database in one API call in order to return 100 leads.

Reply:
This is still a performance dealbreaker that needs to be fixed. Additionally, there is a race condition here - the code here assumes that the last request to the database is the last response from the database which is not always true.

Reply:
Oh wait I misread the code. Looks like you have fixed those things. Will take a closer look at this later.